### PR TITLE
Do not use color for unchanged coverage #777

### PR DESCRIPTION
**Pull request by Amanjain4269**
The previous boolean isPostiveTrend() is changed to int getTrend() and corresponding file changes are made as per the PR - Do not use color for unchanged coverage #777

#### I have added four test cases for the newly created getTrend() method:
- shouldReturnPositiveTrendForLineMetric()
- shouldReturnNegativeTrendForBranchMetric()
- shouldReturnZeroForZeroDelta()
- shouldReturnZeroWhenDeltaIsNotPresent()

### CODE REVIEW

File: plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildAction.java [hunk 1]
```diff
@@ -544,18 +544,29 @@ public String formatDelta(final Baseline baseline, final Metric metric) {
...
     @SuppressWarnings("unused") // Called by jelly view
-    public boolean isPositiveTrend(final Baseline baseline, final Metric metric) {
+    public int getTrend(final Baseline baseline, final Metric metric) {
         var delta = getDelta(baseline, metric);
         if (delta.isPresent()) {
-            if (delta.get().compareTo(Fraction.ZERO) > 0) {
-                return metric.getTendency() == MetricTendency.LARGER_IS_BETTER;
+            double deltaValue = delta.get().doubleValue();
+            // to apply rounding off
+            double roundedDelta = Math.round(deltaValue * 100.0) / 100.0;
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildAction.java:555]: It would be much simpler to use:
If value < 0.1 and value > 0.1 then return 0. else return value. There’s no need to use -1 and +1, any value !=0 would make it.

File: plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java [hunk 2]
```diff
@@ -88,4 +114,32 @@ void shouldCreateViewModel() {
         assertThat(action.getTarget()).extracting(CoverageViewModel::getNode).isSameAs(root);
         assertThat(action.getTarget()).extracting(CoverageViewModel::getOwner).isSameAs(action.getOwner());
     }
+
+    @Test
+    void shouldReturnPositiveTrendForLineMetric() {
+        CoverageBuildAction action = createCoverageBuildActionWithDelta(Metric.LINE);
+        int trend = action.getTrend(Baseline.PROJECT, Metric.LINE);
+        assertThat(trend).isEqualTo(1);
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java:122]: Can you make the delta values visible here? Otherwise it is hard to understand the test. It would also make sense to have in one test the boundaries used that return 0, >0, and <0

**Review by uhafner — CHANGES_REQUESTED** [review]
Thanks! Please also check the CI warnings.

**Comment by Amanjain4269** [comment]
### Thanks for the reviews.
I made the changes in another commit. Can you please check now?

File: plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java [hunk 3]
```diff
@@ -88,4 +120,46 @@ void shouldCreateViewModel() {
         assertThat(action.getTarget()).extracting(CoverageViewModel::getNode).isSameAs(root);
         assertThat(action.getTarget()).extracting(CoverageViewModel::getOwner).isSameAs(action.getOwner());
     }
+
+    @Test
+    void shouldReturnPositiveTrendForLineMetric() {
+        CoverageBuildAction action = createCoverageBuildActionWithDelta();
+        double trend = action.getTrend(Baseline.PROJECT, Metric.LINE);
+        assertThat(trend).isEqualTo(0.3);    // deltaValue = 0.3
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java:128]: Almost: what I actually meant is something like:
```suggestion
        CoverageBuildAction action = createCoverageBuildActionWithDelta(Baseline.PROJECT, Metric.LINE, Fraction.of("0.001");
        assertThat(action.getTrend(Baseline.PROJECT, Metric.LINE)).isPositive();
```

and then 

```
        CoverageBuildAction action = createCoverageBuildActionWithDelta(Baseline.PROJECT, Metric.LINE, Fraction.of("0.0009");
        assertThat(action.getTrend(Baseline.PROJECT, Metric.LINE)).isZero();
```

So move the interesting parts to the focus of the reader.

File: plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildAction.java [hunk 4]
```diff
@@ -544,18 +544,25 @@ public String formatDelta(final Baseline baseline, final Metric metric) {
      * @param metric
      *         the metric to check
      *
-     * @return {@code true} if the trend is positive, {@code false} otherwise
+     * @return {@code roundedDeltaValue} if the trend is positive or negative, {@code 0} otherwise
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildAction.java:547]:
```suggestion
     * @return a positive value if the trend is positive, a negative value if the trend is negative, or {@code 0} if there is no significant change in the trend
```

File: plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildAction.java [hunk 5]
```diff
@@ -544,18 +544,25 @@ public String formatDelta(final Baseline baseline, final Metric metric) {
...
-    public boolean isPositiveTrend(final Baseline baseline, final Metric metric) {
+    public double getTrend(final Baseline baseline, final Metric metric) {
         var delta = getDelta(baseline, metric);
         if (delta.isPresent()) {
-            if (delta.get().compareTo(Fraction.ZERO) > 0) {
-                return metric.getTendency() == MetricTendency.LARGER_IS_BETTER;
+            double deltaValue = delta.get().doubleValue();
+            // to apply rounding off for boundary delta values
+            double roundedDelta = Math.round(deltaValue * 100.0) / 100.0;
+            if (-0.1 < roundedDelta && roundedDelta < 0.1) {
```
Comment (uhafner) [inline plugin/src/main/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildAction.java:556]: Sorry, my fault: percentages are stored in the interval [0,1], so it makes sense to compare 0.001. I also do not see the reason why you are rounding?

File: plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java (hunk 3 above)
Comment (uhafner) [reply plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java:128]: Maybe it would be helpful to add here: 
```
assertThat(action.formatDelta(Baseline.PROJECT, Metric.LINE)).isEqualTo("+0.10%");
```

**Review by uhafner — CHANGES_REQUESTED** [review]
Almost ready for merge. Thanks again!

**Comment by Amanjain4269** [comment]
### I made a new commit. Please check now.
Also, there is one more thing I want to mention. I think there are errors in the changes I made in the coverage-summary.jelly file, as the checks are failing. I tried another approach in this new [commit](https://github.com/jenkinsci/code-coverage-api-plugin/pull/789/files#diff-b88388120cbd77d290657a665d92e08080c25afc24af66e26b382130ef55df3b) but still the same.
Can you help me out with this?

File: plugin/src/main/resources/coverage/coverage-summary.jelly [hunk 6]
```diff
@@ -60,11 +60,14 @@
             <li>${formatter.formatValueWithMetric(value)}
               <j:if test="${it.hasDelta(baseline, value.metric)}">
                 <j:choose>
-                  <j:when test="${it.isPositiveTrend(baseline, value.metric)}">
+                  <j:if test="${it.getTrend(baseline, value.metric) > 0}">
```
Comment (uhafner) [inline plugin/src/main/resources/coverage/coverage-summary.jelly:63]:
```suggestion
                  <j:when test="${it.getTrend(baseline, value.metric) > 0}">
```

File: plugin/src/main/resources/coverage/coverage-summary.jelly [hunk 7]
```diff
@@ -60,11 +60,14 @@
...
                     <j:set var="color" value="var(--green)"/>
-                  </j:when>
-                  <j:otherwise>
+                  </j:if>
+                  <j:if test="${it.getTrend(baseline, value.metric) < 0}">
```
Comment (uhafner) [inline plugin/src/main/resources/coverage/coverage-summary.jelly:66]:
```suggestion
                  <j:when test="${it.getTrend(baseline, value.metric) < 0}">
```

File: plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java [hunk 8]
```diff
@@ -88,4 +107,22 @@ void shouldCreateViewModel() {
...
+    void shouldReturnZeroForDeltaWithinBoundaries() {
+        CoverageBuildAction action = createCoverageBuildActionWithDelta(Baseline.PROJECT, Metric.LINE, Optional.of(Fraction.getFraction(9,10000)));
+        assertThat(action.getTrend(Baseline.PROJECT, Metric.LINE)).isZero();
+    }
+
+    @Test
+    void shouldReturnZeroWhenDeltaIsNotPresentForGivenMetric() {
+        CoverageBuildAction action = createCoverageBuildActionWithDelta(Baseline.PROJECT, Metric.LINE, Optional.empty());
+        assertThat(action.getTrend(Baseline.PROJECT, Metric.LINE)).isZero();
+    }
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java:127]: Tests look good now!

Just a couple of analysis warnings remaining...

**Review by uhafner — APPROVED** [review]

**Comment by Amanjain4269** [comment]
@uhafner 
### Thanks for your valuable insights.
I made the final changes.

File: plugin/src/main/resources/coverage/coverage-summary.jelly [hunk 9]
```diff
@@ -60,11 +60,14 @@
             <li>${formatter.formatValueWithMetric(value)}
               <j:if test="${it.hasDelta(baseline, value.metric)}">
                 <j:choose>
-                  <j:when test="${it.isPositiveTrend(baseline, value.metric)}">
+                  <j:when test="${it.getTrend(baseline, value.metric) > 0}">
```
Comment (uhafner) [inline plugin/src/main/resources/coverage/coverage-summary.jelly:63]:
```suggestion
                  <j:when test="${it.getTrend(baseline, value.metric) gt 0}">
```

File: plugin/src/main/resources/coverage/coverage-summary.jelly [hunk 10]
```diff
@@ -60,11 +60,14 @@
...
                     <j:set var="color" value="var(--green)"/>
                   </j:when>
-                  <j:otherwise>
+                  <j:when test="${it.getTrend(baseline, value.metric) < 0}">
```
Comment (uhafner) [inline plugin/src/main/resources/coverage/coverage-summary.jelly:66]:
```suggestion
                  <j:when test="${it.getTrend(baseline, value.metric) lt 0}">
```

File: plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java [hunk 11]
```diff
@@ -30,6 +31,25 @@
  */
 @DefaultLocale("en")
 class CoverageBuildActionTest {
+    @SuppressWarnings("unused")
+    private CoverageBuildAction createCoverageBuildActionWithDelta(final Baseline baseline, final Metric metric, final Optional<Fraction> delta) {
```
Comment (uhafner) [inline plugin/src/test/java/io/jenkins/plugins/coverage/metrics/steps/CoverageBuildActionTest.java:35]: Seems that baseline is not used at all:
```suggestion
    private CoverageBuildAction createCoverageBuildActionWithDelta(final Metric metric, final Optional<Fraction> delta) {
```

**Review by uhafner — CHANGES_REQUESTED** [review]
There is a compile error in the Jelly file.

**Comment by Amanjain4269** [comment]
@uhafner  
Is everything OK now?

**Review by uhafner — APPROVED** [review]

**Comment by uhafner** [comment]
Thanks for your patience!
